### ISSUE TITLE
Scripts: Enhance check-engines to consider "latest" versions

### ISSUE THREAD

**Issue by aduth** [WordPress/gutenberg#14201]
Previously: #12721
Related: https://github.com/WordPress/gutenberg/pull/14192#issuecomment-468732835

Currently, `wp-scripts check-engines` is hard-coded to default to Node `>=10.0.0` and npm `>=6.0.0`:

https://github.com/WordPress/gutenberg/blob/62f81e10f92a1ec2bdd71c4fe0e249c1e5cb8ccd/packages/scripts/scripts/check-engines.js#L23-L24

This doesn't faithfully represent the versions Gutenberg expects, which are respectively "latest Node LTS" and "latest npm version"

>You should be running a Node version matching the current active LTS release or newer for this plugin to work correctly.

>You also should have the latest release of npm installed.

https://github.com/WordPress/gutenberg/blob/master/CONTRIBUTING.md#getting-started

If the `check-engines` script is intended to adhere to this recommendation, it should ideally do so in a way which doesn't require manual updates.

It does not appear that the underlying [`check-node-version` project](https://github.com/parshap/check-node-version) supports this, but perhaps part of the script can be to retrieve the latest versions information.

- For npm, the `data[ 'dist-tags' ].latest` result from https://registry.npmjs.org/npm JSON blob
- For Node, the result of `data`, first `version` of sorted by date, filtered by `lts` value equal to the [current LTS codename](https://github.com/nodejs/Release) (currently "Dubnium", may need to be manually updated) from https://nodejs.org/dist/index.json

cc @gziolo @ntwb @adamsilverstein

**Comment by ntwb** [comment]
I wonder if this could be added as an enhancement to the existing `valid-values-engines` rule of `npm-package-json-lint`?

https://github.com/tclindner/npm-package-json-lint/wiki/valid-values-engines

**Comment by gziolo** [comment]
I like this proposal. It would be very helpful. The very naive implementation we have at the moment was very easy to implement. It runs as `predev` and `prebuild` scripts. My only concern would be how it is handled when someone is offline. At the moment it's a non-issue as those numbers are hardcoded. We would need to ensure it has some cached value as a fallback to be used in a scenario like that.

**Comment by aduth** [comment]
> My only concern would be how it is handled when someone is offline. At the moment it's a non-issue as those numbers are hardcoded. We would need to ensure it has some cached value as a fallback to be used in a scenario like that.

That's true. Another thought which had occurred to me is that we set something up which acts similar to [Greenkeeper](https://greenkeeper.io/) in automating pull requests to bump the version once a new one becomes available. It just seemed a bit cumbersome / more involved to implement this way.

> I wonder if this could be added as an enhancement to the existing `valid-values-engines` rule of `npm-package-json-lint`?

How then would it be captured? Is the idea that `check-engines` script would be updated to compare against the `engines` entry? Seems to me like it should be doing that anyways.

> * For Node, the result of `data`, first `version` of sorted by date, filtered by `lts` value equal to the [current LTS codename](https://github.com/nodejs/Release) (currently "Dubnium", may need to be manually updated) from https://nodejs.org/dist/index.json

Update to my own comment: If we wanted this fully automated, the Node Release WG does publish a JSON file with the current LTS codenames:

https://github.com/nodejs/Release/blob/eb91c94/schedule.json#L46-L52

**Comment by gziolo** [comment]
As of https://github.com/WordPress/gutenberg/pull/28143, `check-engines` uses the engines config from `@wordpress/scripts` by default.

It still would be great to implement the solution that would allow to pass a flag like `—latest` that is able to detect proper versions.
